### ISSUE TITLE
org.jruby.rack.RackInitializationException: no such file to load -- bundler/setup

### ISSUE THREAD

**Issue by c0mrade** [jruby/warbler#308]
Hello,

I've tried to automate my deployment steps, including creating a tomcat instance from existing template, creating a war file, uploading assets to CDN, copy war over to webapps etc etc.

Last step is to start the server. This task looks like this :

```
desc 'Start tomcat server'
  task :start_server, [:server_path] => :environment do |t, args|
    server_path = args[:server_path]
    sh "#{server_path}/bin/catalina.sh start"
  end
```

So when I start my server from rake task and open a server url locally this is the error that I get :

```
org.jruby.rack.RackInitializationException: no such file to load -- bundler/setup
    from org/jruby/RubyKernel.java:1085:in `require'
    from file:/apps/tomcat/webapps/ROOT/WEB-INF/lib/jruby-stdlib-complete-1.7.11.jar!/META-INF/jruby.home/lib/ruby/shared/rubygems/core_ext/kernel_require.rb:55:in `require'

    at org.jruby.rack.RackInitializationException.wrap(RackInitializationException.java:29)
    at org.jruby.rack.RackApplicationFactoryDecorator.init(RackApplicationFactoryDecorator.java:104)
    at org.jruby.rack.RackServletContextListener.contextInitialized(RackServletContextListener.java:50)
    at org.apache.catalina.core.StandardContext.listenerStart(StandardContext.java:4772)
    at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5196)
    at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:150)
    at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:725)
    at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:701)
    at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:714)
    at org.apache.catalina.startup.HostConfig.deployWAR(HostConfig.java:917)
    at org.apache.catalina.startup.HostConfig$DeployWar.run(HostConfig.java:1701)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
    at java.util.concurrent.FutureTask.run(FutureTask.java:262)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
    at java.lang.Thread.run(Thread.java:745)
Caused by: org.jruby.exceptions.RaiseException: (LoadError) no such file to load -- bundler/setup
    at org.jruby.RubyKernel.require(org/jruby/RubyKernel.java:1085)
    at RUBY.require(file:/apps/tomcat/webapps/ROOT/WEB-INF/lib/jruby-stdlib-complete-1.7.11.jar!/META-INF/jruby.home/lib/ruby/shared/rubygems/core_ext/kernel_require.rb:55)
```

However when I start a server from my console like this :

```
/apps/tomcat/bin/catalina.sh start
```

Runs without problem. And I've tried to get to the root cause of this, it appears that there is a difference in some Environment variables when starting with rake and from shell.

Anyone had similar issue, and found a solution or more information?
##

**Comment by kares** [comment]
it's hard to tell exactly but there's probably something along side seeing the packed GEM_HOME files

first thing I would try is use the latest JRuby (it's for sure an older 1.7.x version) + JRuby-Rack ...

**Comment by dekz** [comment]
I can reproduce this running Tomcat 8 inside a docker container and having the WAR installed inside the container.

The WAR works fine as a JETTY executable, but inside Tomcat I get this exception.

**Comment by mkristian** [comment]
@dekz did you use jruby-1.7.20 ?

any chance to run tomcat with -Djruby.debug.loadService=true and see the console log ?

**Comment by dekz** [comment]
I have been using this on jruby 9.0.0.0 with the warbler 2.0 branch.

**Comment by dekz** [comment]
[Here is that debug log](https://gist.github.com/dekz/cbb815da9e2c5491ce84)  for the loadService.

**Comment by mkristian** [comment]
thanx for log - does not reveal anything :(

**Comment by mkristian** [comment]
I just run into "a" bundler/setup issue as well in a different context of runnable jars. maybe related - still investigating

**Comment by mkristian** [comment]
so my problem was/is that bundler/setup tries to update Gemfile.lock since the gems found on GEM_PATH/GEM_HOME do not match but altogether are valid resolution and then bundler wants to write out Gemfile.lock with those found versions. BUT this file is inside a jar - in my case - and this fails with NPE

maybe this is related to this issue here

**Comment by kares** [comment]
@mkristian in that case [`BUNDLE_FROZEN`](https://github.com/jruby/warbler/blob/master/lib/warbler/templates/bundler.erb) should be on by default, right?

**Comment by mkristian** [comment]
I think BUNDLE_FROZEN would have given me a better error message to find the root cause of my problem.the NPE is bug on JRuby but still trying to get to smaller test case.

if @dekz is willing to see if bundle frozen helps with his initial setup on tomcat that would be great. I still far away from even having tomcat on my machine right now

**Comment by kares** [comment]
let's call this one resolved -> its a 1.7 EOL report anyways and we have BUNDLE_FROZEN in place
